crates/driver/src/domain/competition/settled/mod.rs

```diff
@@ -0,0 +1,268 @@
+                    },
+                    {
+                        // Convert the executed amount to surplus token so it can be compared with
+                        // the surplus
+                        let executed_in_surplus_token = match self.side {
+                            Side::Sell => {
+                                self.executed.0 * self.prices.custom.sell / self.prices.custom.buy
+                            }
+                            Side::Buy => {
+                                self.executed.0 * self.prices.custom.buy / self.prices.custom.sell
+                            }
+                        };
+                        apply_factor(
+                            executed_in_surplus_token,
+                            match self.side {
+                                Side::Sell => max_volume_factor / (1.0 - max_volume_factor),
+                                Side::Buy => max_volume_factor / (1.0 + max_volume_factor),
+                            },
+                        )?
+                    },
+                )
+                .into(),
+            }),
+            order::FeePolicy::PriceImprovement {
+                factor: _,
+                max_volume_factor: _,
+                quote: _,
+            } => todo!(),
+            order::FeePolicy::Volume { factor } => Some(eth::Asset {
+                token: match self.side {
+                    Side::Sell => self.buy.token,
+                    Side::Buy => self.sell.token,
+                },
+                amount: {
+                    // Convert the executed amount to surplus token so it can be compared with
+                    // the surplus
+                    let executed_in_surplus_token = match self.side {
+                        Side::Sell => {
+                            self.executed.0 * self.prices.custom.sell / self.prices.custom.buy
+                        }
+                        Side::Buy => {
+                            self.executed.0 * self.prices.custom.buy / self.prices.custom.sell
+                        }
+                    };
+                    apply_factor(
+                        executed_in_surplus_token,
+                        match self.side {
+                            Side::Sell => factor / (1.0 - factor),
+                            Side::Buy => factor / (1.0 + factor),
+                        },
+                    )?
+                }
+                .into(),
+            }),
+        }
+    }
+
+    /// Protocol fee is defined by fee policies attached to the order.
+    ///
+    /// Denominated in NATIVE token
+    fn native_protocol_fee(
+        &self,
+        prices: &HashMap<eth::TokenAddress, auction::NormalizedPrice>,
+    ) -> Option<eth::TokenAmount> {
+        big_rational_to_u256(
+            &(self.protocol_fee()?.amount.0.to_big_rational() * self.surplus_token_price(prices).0),
+        )
+        .map(Into::into)
+        .ok()
+    }
```

Comment:
This is fine for the currently implemented volume fee in surplus token. If we go with a sell token based fee, this part would need to be revised. E.g. by adding a `protocol_fee_token_price` functionality.